### PR TITLE
Sort actions by tx sequence when available

### PR DESCRIPTION
**Pull request by 45930**
Relates to https://github.com/o1-labs/o1js/issues/1872

The archive node currently in production does not expose enough information to accurately sort actions and produce an actions hash.

https://github.com/o1-labs/Archive-Node-API/pull/107 is the PR to fix the archive node API, and this is the PR to update the query and sorting in o1js.

### CODE REVIEW

File: src/lib/mina/fetch.ts [hunk 1]
```diff
@@ -710,7 +710,15 @@ async function fetchActions(
     graphqlEndpoint,
     networkConfig.archiveFallbackEndpoints
   );
-  if (error) throw Error(error.statusText);
+  if (error) {
+    // retry once without querying transaction info
+    [response, error] = await makeGraphqlRequest<ActionQueryResponse>(
```
Comment (45930) [inline src/lib/mina/fetch.ts:715]: Rather than retry the query, perhaps we can set the archive node API version in the Mina.Network settings somewhere.  Then we could make a different query based on which version we believe we're talking to.

For this specific example, after we get Minascan to update their version, the new query will work in nearly every case, but we ought to prepare for the case that different versions of the API could be running.

**Comment by 45930** [comment]
This PR is not to be merged until after Minascan deploys the fixes in https://github.com/o1-labs/Archive-Node-API/pull/107

File: CHANGELOG.md [hunk 2]
```diff
@@ -20,6 +20,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 ### Fixed
 
 - Compiling stuck in the browser for recursive zkprograms https://github.com/o1-labs/o1js/pull/1906
+- Sort order for actions now includes the transaction sequence number and the exact account id sequence
```
Comment (Trivo25) [inline CHANGELOG.md:23]: nit: link to PR

File: src/lib/mina/fetch.unit-test.ts [hunk 3]
```diff
@@ -136,96 +137,1181 @@ describe('Fetch', async (t) => {
...
+            hash: '1589729766029695153975344283092689798747741638003354620355672853210932754595',
+          },
+          {
+            actions: [
+              [
+                '17137397755795687855356639427474789131368991089558570411893673365904353943290',
+                '1',
+              ],
+            ],
+            hash: '10964420428484427410756859799314206378989718180435238943573393516522086219419',
```
Comment (Trivo25) [inline src/lib/mina/fetch.unit-test.ts:227]: given our history with the action sequence, I think we should seriously consider some advanced testing here - similar to the property testing we do in other places of the codebase

File: src/lib/mina/fetch.unit-test.ts [hunk 4]
```diff
@@ -136,96 +137,1181 @@ describe('Fetch', async (t) => {
...
+                '7',
+                '5706376253943510402705579272467768662733130021399469094173930489992680696927',
+                '1',
+                '1',
+              ],
+            ],
+            hash: '3432500319723012006583645985859849032413742507994268064744704124629760504866',
+          },
+        ]);
+      });
```
Comment (Trivo25) [inline src/lib/mina/fetch.unit-test.ts:1314]: this is very long and makes it very hard to read, can we clean this up a little?

File: src/lib/mina/fixtures/fetch-actions-response-with-transaction-info.ts [hunk 5]
```diff
@@ -0,0 +1,1864 @@
+export { mockFetchActionsResponse };
```
Comment (Trivo25) [inline src/lib/mina/fixtures/fetch-actions-response-with-transaction-info.ts:1]: see what i wrote above; this could save us a few thousand lines here too

File: src/lib/mina/graphql.ts [hunk 6]
```diff
@@ -317,7 +321,8 @@ const getActionsQuery = (
   publicKey: string,
   actionStates: ActionStatesStringified,
   tokenId: string,
-  _filterOptions?: EventActionFilterOptions
+  _filterOptions?: EventActionFilterOptions,
+  retryWithoutTxInfo: boolean = false
```
Comment (Trivo25) [inline src/lib/mina/graphql.ts:325]: nit: we could generalize this into the options parameter above

File: src/lib/mina/fetch.unit-test.ts (hunk 3 above)
Comment (45930) [reply src/lib/mina/fetch.unit-test.ts:227]: Can you describe what cases you'd like to see?  The existing tests here are simply real payloads from the graphQL node, and the real output of `fetchActions` in a case where the actions are derived successfully.  In particular, it is one of the edge cases discovered by the original reporter of the issue: https://github.com/o1-labs/o1js/issues/1872#issue-2595537724

File: src/lib/mina/fetch.unit-test.ts (hunk 4 above)
Comment (45930) [reply src/lib/mina/fetch.unit-test.ts:1314]: I don't know what you mean by clean it up?  Do you mean move the payload to a JSON file and read it in?

I would push back on removing this test or trying to prune it to its most useful bits.  Since it is a raw input-output from a real edge case, I think it's worth leaving in full.  It's not really meant to be readable, but it will flag regressions.

File: src/lib/mina/fixtures/fetch-actions-response-with-transaction-info.ts (hunk 5 above)
Comment (45930) [reply src/lib/mina/fixtures/fetch-actions-response-with-transaction-info.ts:1]: FWIW, this is the block in which the actions are out of order by the currently-live algorithm.  So if we did trim it, we should at least keep this.

```
      {
        blockInfo: {
          distanceFromMaxBlockHeight: 10037,
        },
        actionState: {
          actionStateOne:
            '11553781833537412187998791170094246504179739906426596650283148421062134031747',
          actionStateTwo:
            '2657110063538905863496230550574434206907731812034404344231638673775478011856',
        },
        actionData: [
          {
            accountUpdateId: '40618',
            data: [
              '7',
              '7',
              '7',
              '7',
              '7',
              '7',
              '15661593045172372274349484602975176160403678593916071611247487060653084577990',
              '0',
              '1',
            ],
            transactionInfo: {
              sequenceNumber: 4,
              zkappAccountUpdateIds: [40618, 6396, 33066],
            },
          },
          {
            accountUpdateId: '40617',
            data: [
              '3',
              '4',
              '5',
              '7',
              '6',
              '2',
              '6453147199940115032546911372596674482366995380739961315278593325733504245402',
              '0',
              '1',
            ],
            transactionInfo: {
              sequenceNumber: 5,
              zkappAccountUpdateIds: [40617, 40372, 40373],
            },
          },
          {
            accountUpdateId: '40619',
            data: [
              '3',
              '4',
              '5',
              '7',
              '6',
              '3',
              '21594276700301685638451290563121567472719745179788000956082286147863718718533',
              '1',
              '1',
            ],
            transactionInfo: {
              sequenceNumber: 6,
              zkappAccountUpdateIds: [40619, 33539, 33540],
            },
          },
          {
            accountUpdateId: '40620',
            data: [
              '3',
              '4',
              '5',
              '7',
              '6',
              '3',
              '21594276700301685638451290563121567472719745179788000956082286147863718718533',
              '1',
              '1',
            ],
            transactionInfo: {
              sequenceNumber: 7,
              zkappAccountUpdateIds: [40620, 33539, 33540],
            },
          },
        ],
      },
      ```

File: src/lib/mina/fetch.unit-test.ts (hunk 4 above)
Comment (Trivo25) [reply src/lib/mina/fetch.unit-test.ts:1314]: I don't want to remove it but I would like to make it more readable (dedicated JSON file for the payloads, for example)

File: src/lib/mina/fixtures/fetch-actions-response-with-transaction-info.ts (hunk 5 above)
Comment (Trivo25) [reply src/lib/mina/fixtures/fetch-actions-response-with-transaction-info.ts:1]: Sorry, I meant the property testing would be benefical here too

File: src/lib/mina/fetch.unit-test.ts (hunk 3 above)
Comment (Trivo25) [reply src/lib/mina/fetch.unit-test.ts:227]: Take a look at this https://github.com/o1-labs/o1js/blob/00eb12df2b35cd7bc6f020570e20c5e28eee8ed5/src/lib/testing/property.ts

File: src/lib/mina/fetch.unit-test.ts (hunk 4 above)
Comment (45930) [reply src/lib/mina/fetch.unit-test.ts:183]: I updated the pattern here, instead of doing a strict `bigObject.equals(anotherObject)`, I just check the hashes, which takes up less room and still inherently tests that the data is in the right order.

File: src/lib/mina/graphql.ts (hunk 6 above)
Comment (45930) [reply src/lib/mina/graphql.ts:325]: `EventActionFilterOptions` directly maps to the graphQL schema, but `retryWithoutTxInfo` is strictly o1js logic, so I think it's ok as is.

I wonder what you think about just skipping this retry logic though.  Minascan has updated their endpoint already, and if anyone else is using a self-hosted archive node API, they will eventually realize they need to upgrade.

I think we need to keep this retry step for a few minor versions at least, just in case Minascan needs to roll back the changes, we would rather o1js continue to work.  But eventually we can just remove this option.

File: src/lib/mina/fetch.ts [hunk 7]
```diff
@@ -757,9 +796,33 @@ export function createActionsList(
         `No action data was found for the account ${publicKey} with the latest action state ${actionState}`
       );
 
-    actionData = actionData.sort((a1, a2) => {
-      return Number(a1.accountUpdateId) < Number(a2.accountUpdateId) ? -1 : 1;
-    });
+    // DEPRECATED: In case the archive node is running an out-of-date version, best guess is to sort by the account update id
+    // As of 2025-01-07, minascan is running a version of the node which supports `sequenceNumber` and `zkappAccountUpdateIds` fields
+    // We could consider removing this fallback since no other nodes are widely used
```
Comment (Trivo25) [inline src/lib/mina/fetch.ts:801]: > We could consider removing this fallback since no other nodes are widely used

I would be careful with that and I don't think we should. App developers are currently incentivized to run their own services for that type of stuff, so we will never know what users are exactly using. Imo we should just keep it in there, it doesn't hurt

**Review by Trivo25 — APPROVED** [review]

File: src/lib/mina/fetch.ts [hunk 8]
```diff
@@ -691,10 +691,30 @@ async function fetchEvents(
...
+ * @param accountInfo - An {@link ActionsQueryInputs} containing the public key, and optional query parameters for the actions query
+ * @param graphqlEndpoint - The GraphQL endpoint to fetch from. Defaults to the configured Mina endpoint.
+ *
+ * @returns A promise that resolves to an object containing the final actions hash for the account, and a list of actions
+ * @throws Will throw an error if the GraphQL endpoint is invalid or if the fetch request fails.
+ *
+ * @example
+ * const accountInfo = { publicKey: 'B62qiwmXrWn7Cok5VhhB3KvCwyZ7NHHstFGbiU5n7m8s2RqqNW1p1wF' };
+ * const acitonsList = await fetchAccount(accountInfo);
+ * console.log(acitonsList);
```
Comment (boray) [inline src/lib/mina/fetch.ts:706]:
```suggestion
 * const actionsList = await fetchAccount(accountInfo);
 * console.log(actionsList);
```
nit: typo

**Review by boray — APPROVED** [review]

File: src/lib/mina/fetch.unit-test.ts (hunk 3 above)
Comment (mitschabaude) [reply src/lib/mina/fetch.unit-test.ts]: yeah I think the problem with property testing (or any unit testing) here is that the e2e lifecycle of actions doesn't live in this repo -- most of it lives in the archive node & API. so a unit test of correct behavior given arbitrary inputs would make sense there, but not here

File: src/lib/mina/fetch.ts (hunk 7 above)
Comment (45930) [reply src/lib/mina/fetch.ts:801]: > Imo we should just keep it in there, it doesn't hurt

It's a little sloppy because we will retry on _any_ error.  The archive node API responds with `Error: Unknown Error: {}` when this situation is hit.  So we should also improve the error handling on the API side.

But I'll leave it as-is for now.  The most common case will be that that the first request succeeds and this case is never hit.
